modules/apps/repository/repository-test/src/testIntegration/java/com/liferay/repository/test/RepositoryFactoryDBPartitionTest.java

```diff
@@ -0,0 +1,215 @@
+import com.liferay.portal.kernel.security.auth.CompanyThreadLocal;
+import com.liferay.portal.kernel.security.auth.PrincipalThreadLocal;
+import com.liferay.portal.kernel.service.GroupLocalService;
+import com.liferay.portal.kernel.service.RepositoryLocalService;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.service.UserLocalService;
+import com.liferay.portal.kernel.test.util.RandomTestUtil;
+import com.liferay.portal.kernel.util.FriendlyURLNormalizerUtil;
+import com.liferay.portal.kernel.util.HashMapBuilder;
+import com.liferay.portal.kernel.util.LocaleUtil;
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.kernel.util.UnicodeProperties;
+import com.liferay.portal.repository.liferayrepository.LiferayRepository;
+import com.liferay.portal.test.rule.Inject;
+
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Marco Galluzzi
+ */
+@RunWith(Arquillian.class)
+public class RepositoryFactoryDBPartitionTest extends BaseDBPartitionTestCase {
+
+	@BeforeClass
+	public static void setUpClass() throws Exception {
+		BaseDBPartitionTestCase.setUpClass();
+
+		BaseDBPartitionTestCase.setUpDBPartitions();
+	}
+
+	@AfterClass
+	public static void tearDownClass() throws Exception {
+		BaseDBPartitionTestCase.tearDownDBPartitions();
+	}
+
+	@Test
+	public void testAddFolderInRepositoryWhenSameRepositoryIdExistsInDifferentPartition()
+		throws Exception {
+
+		IndexStatusManagerThreadLocal.setIndexReadOnly(true);
+
+		long counterValue = _getCounterValue();
+
+		long companyId = COMPANY_IDS[0];
+		long groupId = counterValue + 100;
+		long repositoryId = counterValue + 1000;
+
+		try (SafeCloseable safeCloseable =
+				CompanyThreadLocal.setWithSafeCloseable(companyId)) {
+
+			User user = _userLocalService.getGuestUser(companyId);
+
+			Group group = _addGroup(user.getUserId(), groupId);
+
+			Assert.assertEquals(groupId, group.getGroupId());
+
+			com.liferay.portal.kernel.model.Repository repositoryModel =
+				_addRepository(group, user.getUserId(), repositoryId);
+
+			Assert.assertEquals(
+				repositoryId, repositoryModel.getRepositoryId());
+
+			_repositoryFactory.createRepository(
+				repositoryModel.getRepositoryId());
+		}
+
+		companyId = COMPANY_IDS[1];
+		groupId = counterValue + 200;
+
+		try (SafeCloseable safeCloseable =
+				CompanyThreadLocal.setWithSafeCloseable(companyId)) {
+
+			User user = _userLocalService.getGuestUser(companyId);
+
+			Group group = _addGroup(user.getUserId(), groupId);
+
+			Assert.assertEquals(groupId, group.getGroupId());
+
+			com.liferay.portal.kernel.model.Repository repositoryModel =
+				_addRepository(group, user.getUserId(), repositoryId);
+
+			Assert.assertEquals(
+				repositoryId, repositoryModel.getRepositoryId());
+
+			Repository repository = _repositoryFactory.createRepository(
+				repositoryModel.getRepositoryId());
```

Comment:
This piece of code is a literal duplication of the one above, and here you're mixing setup and assertion code. Can you refactor this so that we have three different pieces and also reduce duplication?

Reply:
Refactored